src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java

```diff
@@ -1073,4 +1073,34 @@ public void deleteTasksWithSolutionEntries(Long exerciseId) {
         programmingExerciseTaskRepository.deleteAll(tasks);
         programmingExerciseSolutionEntryRepository.deleteAll(solutionEntries);
     }
+
+    /**
+     * Locks or unlocks the repository if necessary due to the changes in the programming exercise
+     * @param programmingExerciseBeforeUpdate the original exercise with unchanged values
+     * @param updatedProgrammingExercise the updated exercise with new values
+     */
+    public void handleRepoAccessRightChanges(ProgrammingExercise programmingExerciseBeforeUpdate, ProgrammingExercise updatedProgrammingExercise) {
+        ZonedDateTime now = ZonedDateTime.now();
+        if (updatedProgrammingExercise.getDueDate() == null || updatedProgrammingExercise.getDueDate().isAfter(now)) {
+            if (Boolean.FALSE.equals(programmingExerciseBeforeUpdate.isAllowOfflineIde()) && Boolean.TRUE.equals(updatedProgrammingExercise.isAllowOfflineIde())) {
+                unlockAllRepositories(programmingExerciseBeforeUpdate.getId());
+            }
+            else if (Boolean.TRUE.equals(programmingExerciseBeforeUpdate.isAllowOfflineIde()) && Boolean.FALSE.equals(updatedProgrammingExercise.isAllowOfflineIde())) {
+                lockAllRepositories(programmingExerciseBeforeUpdate.getId());
+            }
+        }
+
+        if (updatedProgrammingExercise.isAllowOfflineIde()) {
+            if (programmingExerciseBeforeUpdate.getDueDate() != null && programmingExerciseBeforeUpdate.getDueDate().isBefore(now)
+                    && (updatedProgrammingExercise.getDueDate() == null || updatedProgrammingExercise.getDueDate().isAfter(now))) {
+                // New due date allows students to continue working on exercise
+                unlockAllRepositories(programmingExerciseBeforeUpdate.getId());
+            }
+            else if ((programmingExerciseBeforeUpdate.getDueDate() == null || programmingExerciseBeforeUpdate.getDueDate().isAfter(now))
+                    && updatedProgrammingExercise.getDueDate() != null && updatedProgrammingExercise.getDueDate().isBefore(now)) {
+                // New due date forbids students to continue working on exercise
```

Comment:
I think lockAllRepositories does not look into the individual due date, but this should get respected in cases like this,

Reply:
Handled now